functional_programming_tests/test/with_dependency_injection.cpp

```diff
@@ -0,0 +1,528 @@
+              (override));
+  MOCK_METHOD(void, register_generate_path_service,
+              (GeneratePathServiceCallback), (override));
+  MOCK_METHOD(void, log_info, (std::string const&), (override));
+  MOCK_METHOD(void, log_error, (std::string const&), (override));
+};
+
+TEST(PathGenerator, Ctor) {
+  // GIVEN a mock middleware handle
+  auto mw = std::make_unique<MockMiddlewareHandle>();
+
+  // THEN it should register the services
+  EXPECT_CALL(*mw, register_set_map_service(testing::_)).Times(1);
+  EXPECT_CALL(*mw, register_generate_path_service(testing::_)).Times(1);
+
+  // WHEN the path generator is constructed
+  auto const path_generator = PathGenerator{std::move(mw)};
+}
+
+TEST(PathGenerator, SetMap) {
+  // GIVEN a path generator
+  auto mw = std::make_unique<MockMiddlewareHandle>();
+  // Capture the callback so it can be called later
+  PathGenerator::MiddlewareHandle::SetMapServiceCallback callback;
+  ON_CALL(*mw, register_set_map_service(testing::_))
+      .WillByDefault(testing::SaveArg<0>(&callback));
+
+  auto const path_generator = PathGenerator{std::move(mw)};
+
+  // WHEN the set map service is called
+  auto const request = make_costmap();
+  auto response = std::make_shared<example_srvs::srv::SetMap::Response>();
+  callback(request, response);
+
+  // THEN the path generator should successfully set the map
+  EXPECT_TRUE(response->success.data);
+}
+
+TEST(PathGenerator, NoCostmap) {
+  // GIVEN a path generator with a costmap
+  auto mw = std::make_unique<MockMiddlewareHandle>();
+  // Capture the path callback so it can be called later
+  PathGenerator::MiddlewareHandle::GeneratePathServiceCallback path_callback;
+  ON_CALL(*mw, register_generate_path_service(testing::_))
+      .WillByDefault(testing::SaveArg<0>(&path_callback));
+
+  auto const path_generator = PathGenerator{std::move(mw)};
+
+  // WHEN the generate path service is called without a costmap
+  auto path_request = std::make_shared<example_srvs::srv::GetPath::Request>();
+  path_request->start.data = {0, 0};
+  path_request->goal.data = {0, 0};
+
+  auto path_response = std::make_shared<example_srvs::srv::GetPath::Response>();
+  path_callback(path_request, path_response);
+
+  // THEN the path generator should fail
+  EXPECT_FALSE(path_response->success.data);
+}
+
+struct PathGeneratorFixture : public testing::Test {
+  /**
+   * @brief Construct a fixture which will set the map and capture the path
+   * callback
+   */
+  PathGeneratorFixture() : mw_{std::make_unique<MockMiddlewareHandle>()} {
+    // When the map callback is called, set the costmap
+    ON_CALL(*mw_, register_set_map_service(testing::_))
+        .WillByDefault([&](auto const& map_callback) {
+          auto const map_request = make_costmap();
+          auto map_response =
+              std::make_shared<example_srvs::srv::SetMap::Response>();
+          map_callback(map_request, map_response);
+        });
```

Comment:
Does this mean that when `register_set_map_service` is called by in the PathGenerator constructor, the argument to the `register_set_map_service` is input to the lambda here? In this case, `map_callback`, is the lambda on line 108, and this lambda uses that `map_callback` to set the costmap inside of the PathGenerator?

Reply:
Yessir